### DISCUSSION TITLE
Wordstar 4 : not enough memory

### DISCUSSION THREAD

**Discussion by ghost** [thehighnibble/thehighnibble.github.io#69] [Q&A] [answered]
Hello everyone,

In the default install of our beloved IMSAI8080 replica comes mounted on I: an hard disk with Wordstar 4

Launching it (in the default cpm2.2 mode), it's unable to load/create a document (d) complaining of **not enough memory**.
Does this have a solution ?
Opening a NON document does work

Wordstar 3.3 (on a floppy) can do both and doesn't complain about memory

Just a FYI,

**Comment by nbreeden2** [comment]
How much memory does you CP/M build report when started. CP/M reports it on startup on the TTY: before it presents the prompt. I can't test this myself as WS has been deleted from my I: drive to make room for a couple C compilers.

I've updated my boot floppy for 64K RAM.
64K CP/M VERS 2.2 B01
A>

**Comment by ghost** [comment]


**Reply by nbreeden2** [reply, marked as answer]
In the end it really comes down to the Startup Configuration.

@dmcnaugh is correct that in the configuration he is demoing in the video you reference 54K is the limit.

Keep in mind that S-100 platforms such as the IMSAI have highly configurable hardware. Other systems such as the TRS-80, Apple ][ or C64 are pretty much locked down from a hardware view, you can add things such as peripherals or memory BUT you can't move the KERNAL or ROMs to different addresses or change how the keyboard is addressed. In a general S-100 platform none of this is predetermined. 

With that said some manufactures of S-100 systems built there systems with a base configuration in mind and with some parts such as ROM or I/O configured as they deemed best. This way they could ship a prebuilt CPM floppy with the system so the user didn't need to configure CPM on their own.

It you plan to run CPM there are a few things CPM expects the hardware configuration to meet such as RAM in low memory, 20K minimum RAM (more is highly preferred). This is due to base assumptions that are built into the CPM design.

If you plan to run BASIC in ROM these typically have a pretty locked down set of configurations.

Built into the IMSAI emulation are several memory configurations.

Open SYS:; click the gear icon on the top right hand side; a CFG: window opens; select system.conf; scroll down and you will see various memory configurations.

**[MEMORY 1]**
This is the one I use to get 64K for CPM as the ROM (mpu-a-rom.hex) can be bank switched out for RAM.
#68 has a description of the MPU-B band switching as implemented in the emulation.
[Discussion on the MPU-B implementation for Bank Switching](https://github.com/orgs/thehighnibble/discussions/68#discussioncomment-6019567)
**[MEMORY 2]**
With the two ROMs in this configuration then yes, 54K is the limit. 
If you are going to do anything with the VIO then this is the memory map to use.
**[MEMORY 3]**
If you are going tyo use the MEMON80 ROM then this is the config to use.
It looks like you could get 63K in this config but I haven't tried it.
**[MEMORY 4]**
Configuration for running XYBASIC at startup from ROM (CPM isn't needed).
The ROM takes up 16K leaving 48K user RAM.

Setting which memory map is used is done from the Startup Configuration mode (The Configuration Switch Settings).
[How to configure the MEMORY Model](https://thehighnibble.com/imsai8080/configure/#startup-configuration-non-volatile-storage-nvs)

I needed a CPM configuration with 64K of RAM. The MESCC C compiler needs 64K or RAM to be useful. 
This lead to mepicking a memory configuration that would allow 64K of RAM then modifying a CPM boot floppy to support 64K RAM. CPM needs to be configured for the amount of RAM you have. If you have 64K and CPM is configured for 54K then 54K is all you will get.

The steps to do this would be very difficult to present in text. I'm thinking about doing a video on how to build this configuration.

**Comment by ghost** [comment]
Thanks again Niel, really appreciate.  I did not understand that some memory config end up not using the shadow rom system that we discussed "in extenso" in the other thread, and have ROM that occupy space (like in an Apple II). I have much to learn, I realize that.

To be fair, I'm only following the documentation by having CPM2.2 with 54K of ram and I was just pointing out that by following the doc, you end up with a config that cannot run wordstar, which is also default in drive I:
I realize there are more important problems than that in life :) but hey - here it is. There is such a care of detail in all this project that I'm sure Dave will pick this up. He should either recommend to boot from 64K mode by default or put something else than WS4.0 in drive I: by default, imho. Because anyone following the doc like I did and wanting to test what's on drive I: will end up with the "not enough memory" error.

I would by the way suggest to put XYBASIC in drive I: by default, which would be something more people are wanting to try right away, especially as I think WS3.3 is also loaded by default in a virtual floppy.

Indeed in my case, I'm an old terminal nerd - damn I'm even connecting to my IMSAI 8080 thru Telnet - so I'm unlikely to use VIO at all , so it would make sense for me to try that 64K RAM setup.

**Reply by udo-munk** [reply]
For the IMSAI we emulate all and any hardware for which good enough documentation is available. Now that results to lots and lots of different system configurations, but that is how it was with the S100 systems.

The 54K CP/M will run with any hardware configuration, that is why you get this as a default. Some CP/M software will require a larger TPA for running, not just WS4. On the CP/M disk you will find submit scripts that generate 60K and 64K CP/M versions, but theses require different hardware. For example such a system cannot include the VIO, because this card uses memory at 0xdxxx. Or you might want to include some ROM in the system with a monitor program. This would be placed in upper memory at 0xfxxx, so that leaves 60K for CP/M.

The MPU-A and B ROM's were read out of original IMSAI machines, this is the original code that came with the machines. This ROM's were implemented as shadow ROM's and can be switched off to release the full 64K memory for CP/M. So for this you need to use one of these ROM's, for example MEMON/80 doesn't support this feature, the ROM will always be mapped in upper memory.

Unfortunately you'll have to learn all the nasty details for different hardware configurations to make full use of all this. You will have to read through the original IMSAI documentation for understanding the hardware in detail. This will require time and until you are not familiar with everything you can just use the 54K CP/M, which will just work without reading and learning lots of stuff.

**Comment by ghost** [comment]


**Reply by udo-munk** [reply]
Setting 000 is no ROM and 64 KB RAM only. All you can do is put something into RAM with the frontpanel.
Of course you could try entering a bootloader with the frontpanel, then you won't need a ROM to boot some disk based OS.

**Reply by dmcnaugh** [reply]
Referring back to the original post, of which this thread has strayed significantly:
- The copy of WordStar 4 supplied on the sample hard disk file is not configured for any specific machine
- It does come with a readme file `WSREADME.TXT`  that contains a section titled `MEMORY USAGE`
  - this gives instructions on the memory requirements and how to check what you have available 
  - one solution that requires no change to CP/M is to run the `WSCHANGE.COM` utility that comes with WS4 to reduce WordStar's memory requirements
  - I have just tested this and it solves the _out of memory_ problem you have described
- working with CP/M and applications for this platform was rarely an "out-of-the-box" experience
- you need to be prepared to read everything you can find (and that is provided) and be prepared to investigate and experiment
- The next alternative is to generate a copy of CP/M for 60K or 64K of RAM and Udo has provided utility scripts on the CP/M 2.2 disk to automate this see: `SYSGEN60.SUB` and `SYSGEN64.SUB`  on that disk

> So unless I'm missing something, there is no configuration in the stock nvm memory maps that would actually free up 64KB after bootstrapping, right ?

All the information is contained on the page https://thehighnibble.com/imsai8080/configure/ and the very detailed explanations given earlier in this discussion thread.
You should experiment and learn by your own investigations. This is always more rewarding than being given the answers on a plate.
